File: example/learning-agent/java/src/main/java/org/aiddl/example/learning_agent/functions/OperatorExecutor.java
package org.aiddl.example.learning_agent.functions;

import java.util.Map;

import org.aiddl.common.java.planning.PlanningTerm;
import org.aiddl.common.java.planning.state_variable.Operator;
import org.aiddl.core.java.interfaces.ConfigurableFunction;
import org.aiddl.core.java.representation.CollectionTerm;
import org.aiddl.core.java.representation.SetTerm;
import org.aiddl.core.java.representation.Substitution;
import org.aiddl.core.java.representation.Term;
import org.aiddl.core.java.representation.TupleTerm;
import org.aiddl.core.java.function.FunctionRegistry;
import org.aiddl.core.java.tools.Logger;

public class OperatorExecutor implements ConfigurableFunction {
	
	private String name = this.getClass().getSimpleName();
	private boolean verbose = false;
	
	@Override
	public void configure(Map<Term, Term> settings, FunctionRegistry fReg) {
		this.verbose = Term.sym("true").equals(settings.get(Term.sym("verbose")));
	}

	@Override
	public Term apply( Term args ) {
		Term s = args.get(PlanningTerm.State);
		Term a = args.get(Term.sym("action"));
		CollectionTerm O = (CollectionTerm) args.get(PlanningTerm.Operators);
		
		Term s_next = null;
		for ( Term o : O ) {
			Term name = o.get(PlanningTerm.Name);
			Substitution sub = name.match(a);
			if ( sub != null ) {
				TupleTerm o_sub = (TupleTerm) o.substitute(sub);
				Operator svo = new Operator(o_sub);
				s_next = svo.applyTo((SetTerm) s);
				if ( verbose ) {
					Logger.msg(this.name, "Applying: " + name);
					Logger.msg(this.name, "  s  = " + s);
					Logger.msg(this.name, "  s' = " + s_next);
				}
				break;
			}
		}
		
		if ( s_next == null ) {
			if ( verbose ) Logger.msg(name, "Failure: Action " + a + " not applicable to " + s );
			s_next = s;
		} else {		
			if ( verbose ) Logger.msg(name, "Success: " + s + " x " + a + " -> " + s_next);
		}
	
		return s_next;
	}
}

File: example/learning-agent-scala/src/main/scala/org/aiddl/example/learning_agent_scala/function/OperatorExecutor.scala
package org.aiddl.example.learning_agent_scala.function

import org.aiddl.core.scala.function.{Function, Verbose}
import org.aiddl.core.scala.representation.{CollectionTerm, KeyVal, ListTerm, SetTerm, Sym, Term, Tuple}
import org.aiddl.common.scala.planning.PlanningTerm
import org.aiddl.common.scala.planning.state_variable.{ApplicableFunction, StateTransition}

class OperatorExecutor(operators: CollectionTerm) extends Function with Verbose {
  private val fApplyOperator = new StateTransition()

  override def apply(x: Term): Term = {
    val s = x(PlanningTerm.State).asSet
    val a = x(Sym("action"))
    this.apply(s, a)
  }

  def apply(state: SetTerm, action: Term): SetTerm = {
    var s_next: Term = Sym("NIL")
    for (o <- this.operators) {
      val name = o(PlanningTerm.Name)
      name.unify(action) match {
        case Some(sub) => {
          val o_sub = (o \ sub).asTup
          s_next = fApplyOperator(o_sub, state.asSet)

          this.logger.info("Applying: " + name)
          this.logger.info("  s  = " + state)
          this.logger.info("  s' = " + s_next)
        }
        case None => {}
      }
    }

    if (s_next == Sym("NIL")) {
      this.logger.info("Failure: Action " + action + " not applicable to " + state)
      s_next = state
    } else {
      this.logger.info("Success: " + state + " x " + action + " -> " + s_next)
    }
    s_next.asSet
  }
}
